src/persist/src/indexed/mod.rs

```diff
@@ -1308,6 +1469,254 @@ mod tests {
+        block_on_drain(&mut i, |i, handle| {
+            i.allow_compaction(vec![(id, Antichain::from_elem(5))], handle)
+        })?;
+
+        Ok((i, maintainer, id))
+    }
+
+    /// Test the communication protocol between [Indexed] and [Maintainer]
+    /// specifically for trace compaction requests.
+    #[test]
+    fn trace_compaction() -> Result<(), Error> {
+        let mut unreliable = UnreliableHandle::default();
+        let (mut i, maintainer, id) = trace_compaction_setup(unreliable.clone())?;
+
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request = reqs[0].clone();
+
+        // Once a request is in flight, don't send requests for the same stream.
+        let reqs = i.step()?;
+        assert_eq!(reqs, vec![]);
+
+        // Handle receiving an error MaintenanceRes.
+        let error_response =
+            MaintenanceRes::CompactTrace((id, Err(Error::from("test compaction error"))));
+        i.handle_maintenance_res(error_response.clone());
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // After receiving an error, retry the original request.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request);
+
+        // Handle receiving a valid MaintenanceRes.
+        let response =
+            futures_executor::block_on(request.clone().to_future(&maintainer).run_async());
+
+        i.handle_maintenance_res(response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // After receiving a valid response, issue a new compaction request.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request2 = reqs[0].clone();
+        assert!(request2 != request);
+
+        // Construct a new valid, response.
+        let response2 =
+            futures_executor::block_on(request2.clone().to_future(&maintainer).run_async());
+
+        // Send back an error response. After this there are no requests in flight.
+        i.handle_maintenance_res(error_response);
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // Retry request after error response.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request2);
+
+        // Handle failed write to metadata after receiving a response.
+        i.handle_maintenance_res(response2.clone());
+        unreliable.make_unavailable();
+        assert_eq!(i.drain_pending(), Err(Error::from("unavailable: blob set")));
+        unreliable.make_available();
+
+        // Retry request after failed meta write.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request2);
+
+        // Handle receiving a response after the trace is deleted.
+        block_on(|res| i.destroy("0", res))?;
+
+        // Handle receiving an valid MaintenanceRes for a stream that has been deleted.
+        i.handle_maintenance_res(response2);
+        assert_eq!(i.drain_pending(), Ok(()));
```

Comment:
let's assert on the size of in_flight_trace_compactions here (should be 0, right?)

Reply:
done here and elsewhere

Reply:
i see it elsewhere but not here :). maybe github is lying to me?

Reply:
ah whoops - I added it everywhere in this test:

```
diff --git a/src/persist/src/indexed/mod.rs b/src/persist/src/indexed/mod.rs
index 06bf78fe1..59f599829 100644
--- a/src/persist/src/indexed/mod.rs
+++ b/src/persist/src/indexed/mod.rs
@@ -1728,6 +1728,7 @@ mod tests {
         // After receiving a valid response, issue a new compaction request.
         let reqs = i.step()?;
         assert_eq!(reqs.len(), 1);
+        assert_eq!(i.in_flight_trace_compactions.len(), 1);
         let request2 = reqs[0].clone();
         assert!(request2 != request);

@@ -1737,10 +1738,12 @@ mod tests {
         // Send back an error response. After this there are no requests in flight.
         i.handle_maintenance_res(error_response);
         assert_eq!(i.drain_pending(), Ok(()));
+        assert_eq!(i.in_flight_trace_compactions.len(), 0);

         // Retry request after error response.
         let reqs = i.step()?;
         assert_eq!(reqs.len(), 1);
+        assert_eq!(i.in_flight_trace_compactions.len(), 1);
         assert_eq!(reqs[0], request2);

         // Handle failed write to metadata after receiving a response.
@@ -1748,9 +1751,11 @@ mod tests {
         unreliable.make_unavailable();
         assert_eq!(i.drain_pending(), Err(Error::from("unavailable: blob set")));
         unreliable.make_available();
+        assert_eq!(i.in_flight_trace_compactions.len(), 0);

         // Retry request after failed meta write.
         let reqs = i.step()?;
+        assert_eq!(i.in_flight_trace_compactions.len(), 1);
         assert_eq!(reqs.len(), 1);
         assert_eq!(reqs[0], request2);

@@ -1760,9 +1765,11 @@ mod tests {
         // Handle receiving an valid MaintenanceRes for a stream that has been deleted.
         i.handle_maintenance_res(response2);
         assert_eq!(i.drain_pending(), Ok(()));
+        assert_eq!(i.in_flight_trace_compactions.len(), 0);

         // Don't produce any more requests because the stream is deleted.
         let reqs = i.step()?;
+        assert_eq!(i.in_flight_trace_compactions.len(), 0);
         assert_eq!(reqs, vec![]);
```